### PR TITLE
Implement the Events API

### PR DESCRIPTION
**Pull request by tgoyne**
Not mergeable yet as it depends on https://github.com/realm/realm-core/pull/5436, but the code changes here are reviewable now.

### CODE REVIEW

**Review by leemaguire — COMMENTED** [review]
This is looking pretty good, I left a few small comments. I also notice that there are no Objective-C tests, but given this project's use case and implementation I think just Swift tests should be enough?

File: Realm/RLMAccessor.mm [hunk 1]
```diff
@@ -921,6 +923,11 @@ void RLMSetSwiftPropertyAny(__unsafe_unretained RLMObjectBase *const obj, uint16
 }
 
 id RLMAccessorContext::box(realm::Obj&& r) {
+    if (!currentProperty) {
+        // If currentProperty is set then the caller took care of reporting
+        // the audit event
+        realm::Object(_realm->_realm, *_info.objectSchema, r, _parentObject, _colKey);
```
[the remark this answers is not in the document]
Comment (tgoyne) [reply Realm/RLMAccessor.mm:929]: Creating an `Object` instance reports an audit read. I'll make the comment more descriptive.

File: Realm/RLMAudit.h [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+                                              NSDictionary<NSString *, NSString *> *newMetadata);
+
+@interface RLMAuditConfiguration : NSObject
+@property (nonatomic) NSString *partitionPrefix;
+@property (nonatomic, nullable) RLMUser *syncUser;
+@property (nonatomic, nullable) NSDictionary<NSString *, NSString *> *metadata;
+@property (nonatomic, nullable) void (^logger)(RLMSyncLogLevel, NSString *);
+@property (nonatomic, nullable) void (^errorHandler)(NSError *);
+
+#ifdef __cplusplus
```
Comment (jsflax) [inline Realm/RLMAudit.h:54]: Why is there cpp code exposed in this header?

File: Realm/RLMEvent.h [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+////////////////////////////////////////////////////////////////////////////
+
+#import <Foundation/Foundation.h>
+#import <Realm/RLMConstants.h>
+
+#ifdef __cplusplus
+#include <memory>
+
+namespace realm {
+struct AuditConfig;
```
Comment (bmunkholm) [inline Realm/RLMEvent.h:26]: Shouldn't this be "EventConfig"?

File: Realm/RLMEvent.h [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+@interface RLMEventConfiguration : NSObject
+@property (nonatomic) NSString *partitionPrefix;
+@property (nonatomic, nullable) RLMUser *syncUser;
+@property (nonatomic, nullable) NSDictionary<NSString *, NSString *> *metadata;
+@property (nonatomic, nullable) void (^logger)(RLMSyncLogLevel, NSString *);
+@property (nonatomic, nullable) void (^errorHandler)(NSError *);
+
+#ifdef __cplusplus
+- (std::shared_ptr<realm::AuditConfig>)auditConfigWithRealmConfiguration:(RLMRealmConfiguration *)realmConfig;
```
Comment (bmunkholm) [inline Realm/RLMEvent.h:55]: auditConfigWithRealmConfiguration -> eventConfigWithRealmConfiguration ?

**Review by bmunkholm — COMMENTED** [review]
It looks like the "audit" naming hasn't been completely updated to "event". But also that it's not external facing so likely ok for now. It would be good to rename the internals at some point for internal consistency.

File: Realm/RLMEvent.h (hunk 3 above)
Comment (tgoyne) [reply Realm/RLMEvent.h:26]: No, all of the c++ stuff is still named Audit because that renaming came after it was merged.

**Pull request merged**